tests/jdom/adapters.spec.ts

```diff
@@ -0,0 +1,153 @@
+        const onHandler = (device: JDDevice) => {
+            if (devicesSet.has(device)) {
+                announcedSet.add(device)
+            }
+            
+            if (setEquals(devicesSet, announcedSet)) {
+                bus.off(DEVICE_ANNOUNCE, onHandler)
+                resolve(undefined)
+            }
+        }
+        bus.on(DEVICE_ANNOUNCE, onHandler)
+    })
+
+    // Assign roles now that services are available
+    const roleNameToService: [string, JDService][] = serverDeviceRoleList.filter(([server, device, role]) => {
+        return !!role  // filter for where role name is availab;e
+    }).map(([server, device, role]) => {
+        const services = device.services({serviceClass: server.serviceClass})
+        assert(services.length > 0, `created device ${device.friendlyName} has no service of ${server.specification.name}`)
+        assert(services.length == 1, `created device ${device.friendlyName} has multiple service of ${server.specification.name}`)
+        return [role, services[0]]
+    })
+
+    const roleManager = new RoleManager(bus)
+    roleManager.setRoles(roleNameToService.map(([roleName, service]) => {
+        // TODO role manager doesn't allow manual specificiation of binding role => services right now,
+        // it's greedily / automatically allocated.
+        // When manual specification is added this needs to perform that binding.
+        // For now this just allocates a role and type.
+        // IF YOU HAVE MORE THAN ONE SERVICE INSTANCE WEIRD THINGS CAN HAPPEN!
+        return {role: roleName, serviceShortId: service.specification.shortId}
+    }))
+
+    // TODO this is (might be?) a nasty hack to associate a particular service with a role name
+    roleNameToService.map(([roleName, service]) => {
+        service.role = roleName
+    })
+
+    // TODO HACK HACK HACK
+    // Give adapters a role manager
+    serverDeviceRoleList.forEach(([server, device, role]) => {
+        if (server instanceof AdapterServer) {
+            server._hack_setRoleManager(roleManager)
+        }
+    })
+
+    // Ensure adapters are ready
+    // TODO WRITE ME
+
+    // TODO return devices as map?
+    return bus
+}
+
+suite('adapters', () => {
+    let bus: JDBus;
+
+    afterEach(() => bus?.stop())
+
+    test('click detect event', async function() {
+        console.log("start test")
+
+        // These are here so we have a handle
+        const buttonServer = new ButtonServer("button")  // interface name is just a human-friendly name, not functional
+        const gestureAdapter = new ButtonGestureAdapter("button", "gestureAdapter")
+
+        bus = await createBus([
+            {server: buttonServer, roleName: "button"},
+            {server: gestureAdapter},
+        ])
+        console.log("bus created")
+
+        bus.services({serviceClass: SRV_BUTTON}).forEach(buttonService => {
+            buttonService.on(EVENT, (evs: JDEvent[]) => {  // TODO make sure we're expecting the right events
```

Comment:
argument should really be (event: JDEvent), looks like a bug in my ``emitPropagated`` function. Eventually we'll fix this.